### PR TITLE
Improve handling and exposing of Capabilities

### PR DESCRIPTION
**Pull request by pupnewfster**
## Changes proposed in this pull request:

- Have returned capabilities for a block's side obey the machine's configuration
- Support for "OffsetCapabilities". This system is used by the Digital Miner and the Advanced Bounding Blocks, to make it so that energy cables and item cables can only connect to the proper and marked spots on the miner.
- Makes it a lot easier to disable specific capabilities for a Tile Entity
  - ~Bins now use this to disable their sides from being interacted with instead of having special connection handling~
  - Power accepting tiles only expose the capability on a side that they can receive/give power
  - Multiblocks don't expose the item capability as "randomly" such as it is no longer possible  to just connect a logistical transporter to a reactor casing.

I believe with the way that this allows for removing the `TUBE_CONNECTION_CAPABILITY` all together/simplifying things down to not require having methods like `canTubeConnect` as I believe the reason there were to different capabilities in the first place was so that it can check if it really should have that capability.

Edit: Bins are now able to be interacted with from all sides using logistical transporters.
Edit 2:
Removed a bunch of unchecked cast warnings when it comes to Capabilities, and more importantly removed the need for `TUBE_CONNECTION_CAPABILITY` and use toggleable capabilities instead to only have the `GAS_HANDLER_CAPABILITY` available on some sides. This allows for cleaner implementation of addons wishing to support gas for their machines. On the same note things that previously only implemented/exposed the `ITubeConnection` now support `IGasHandler` so things like the Chemical Oxidizer will work with addons like gas conduits. While nothing relies on or uses `ITubeConnection` or `TUBE_CONNECTION_CAPABILITY` they are still in the code just deprecated so they do not break compat with addons that assume they exist/check for them.

### CODE REVIEW

File: src/main/java/mekanism/common/capabilities/IOffsetCapability.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package mekanism.common.capabilities;
+
+import javax.annotation.Nonnull;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.Vec3i;
+import net.minecraftforge.common.capabilities.Capability;
+
+public interface IOffsetCapability {
```
Comment (thiakil) [inline src/main/java/mekanism/common/capabilities/IOffsetCapability.java]: These interfaces should have some javadocs, for the class and the methods

File: src/main/java/mekanism/common/util/CapabilityUtils.java [hunk 2]
```diff
@@ -21,4 +29,24 @@ public static boolean hasCapability(ICapabilityProvider provider, Capability<?>
 
         return provider.getCapability(cap, side);
     }
+
+    public static <T extends TileEntityContainerBlock & ISideConfiguration> boolean isCapabilityDisabled(
```
Comment (thiakil) [inline src/main/java/mekanism/common/util/CapabilityUtils.java]: This method looks a lot like it should be on `TileComponentConfig` - e.g. make `TileComponentConfig` an `ICapabilityProvider` and delegate in the base class where it is an instance of `ISideConfiguration`

**Comment by thiakil** [comment]
> Bins now use this to disable their sides from being interacted with instead of having special connection handling

Continuing from discord, this will likely break many setups. Historically the ability for Bins to expose `IItemHandler` on the sides is to allow external storage systems to access the contents. E.g. connecting an AE2 Storage Bus to any side.

In 1.7 this was given by the MFR Deep Storage interface, which didn't exist in 1.10 (due to the shift from bundling other people's APIs and MFR not updating for a long time). Thus the special Bin itemhandler implementation was born, to abstract away from the 2 slot internal implementation.

**Comment by pupnewfster** [comment]
> > Bins now use this to disable their sides from being interacted with instead of having special connection handling
> 
> Continuing from discord, this will likely break many setups. Historically the ability for Bins to expose `IItemHandler` on the sides is to allow external storage systems to access the contents. E.g. connecting an AE2 Storage Bus to any side.
> 
> In 1.7 this was given by the MFR Deep Storage interface, which didn't exist in 1.10 (due to the shift from bundling other people's APIs and MFR not updating for a long time). Thus the special Bin itemhandler implementation was born, to abstract away from the 2 slot internal implementation.

I believe the easiest way of doing this will be to just make Bins not be sided, and allow for inserting/extracting from any side including via Logistical Transporters.

**Pull request merged**